### ISSUE TITLE
Improve callback interface for MTK

### ISSUE THREAD

**Issue by AayushSabharwal** [SciML/ModelingToolkit.jl#1734]
Currently, custom callbacks aren't really easy to use with models made using MTK. For example, consider `ContinuousIntegrator`. The `condition` function has the following signature: `condition(u, t, integrator)`. Since MTK generates the function to be solved, the order of variables in `u` is not known to the user. The only real solutions I've found to this problem use a callable struct `function (foo::Foo)(u, t, integrator)` which contains the underlying system, so variables can be accessed. Then, `SciMLBase.sym_to_index` is used to get the index of a symbol in `u`, to actually get the value. Alternatively, store a temporary integrator in the struct which is never solved, but used to access `u` by first calling `set_u!(foo.tmp_integrator, u)` then `foo.tmp_integrator[foo.system.var]`. Both these approaches seem unnecessarily convoluted, especially since I assume `SciMLBase.sym_to_index` isn't meant to be part of the public API.

This is a bit wordy, so if I've done a poor job of explaining the issue please tell me and I'll try to explain it better :)

**Comment by isaacsas** [comment]
Have you looked at https://mtk.sciml.ai/dev/basics/Events/

More generality for conditions than the current symbolic interface would require "generalized conditions", which could be implemented by mimicking what was done for functional affects.

**Comment by AayushSabharwal** [comment]
Thanks! Is that a newer addition? I never noticed it.

**Comment by isaacsas** [comment]
Continuous events have been around for a little while, but the documentation was hidden on a page for component based modeling. Discrete events and functional affects are new, with the dedicated tutorial page just put together in the past week.
